### PR TITLE
BMO CI Decoupling proposal

### PR DESCRIPTION
**Pull request by kashifest**
In this proposal, we discuss about decoupling BMO CI from CAPM3 tests and introduce BMO e2e tests.

### CODE REVIEW

File: design/baremetal-operator/bmo-branching.md
Comment (lentzi90) [inline design/baremetal-operator/bmo-branching.md:21]: Are not the jobs we have for CAPM3 already enough? Or are they not for BMO main branch?

Comment (kashifest) [reply design/baremetal-operator/bmo-branching.md:21]: Actually yes, the main branch test for CAPM3 is already testing with main branch for BMO, I am now thinking would we need similar jobs for CAPM3-BMO release combinations once we have different BMO release branches?

Comment (lentzi90) [reply design/baremetal-operator/bmo-branching.md:21]: I think the simplest would be to lock CAPM3 minor releases to BMO minor releases. So for example the CAPM3 release-1.4 test would run with BMO release-x.y. Then when we release CAPM3 v1.5 we can uplift BMO if there is a new release available and so on.

Comment (tuminoid) [reply design/baremetal-operator/bmo-branching.md:21]: What @lentzi90 said is what we agreed on. We are currently using exact tags of those releases, and we would instead use release branches.

**Comment by kashifest** [comment]
cc @zaneb @honza @dtantsur

**Review by lentzi90 — COMMENTED** [review]
/lgtm

**Comment by tuminoid** [comment]
/hold
until we have more comments (including mine, sorry for delay)

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (Rozzii) [inline design/baremetal-operator/bmo-ci-decoupling.md:10]:
```suggestion
- BMO e2e (completely independent of CAPM3) test has to be designed and put in place. So far BMO PRs are tested with CAPM3, CAPI and IPAM integration tests but with the introduction of BMO release branches and proper releases from those branches, BMO should not rely on CAPM3 releases and should be decoupled from CAPM3 integration tests when testing a BMO PR. We propose a completely independent BMO e2e test framework which will test BMO features with ironic. For this few action points have been identified:
```

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (Rozzii) [inline design/baremetal-operator/bmo-ci-decoupling.md:23]:
```suggestion
- Meanwhile we can branch out BMO to release branch and continue development in main branch. Until BMO e2e tests are in place, we can continue testing BMO release branch and main branch in the same way we do it currently in dev-envs. Instead of using tags which is the way we test currently, we can use branch names for specific branches of CAPM3. For example, CAPM3 release-1.5 branch will be tested with BMO release-0.4 branch and CAPM3 main branch will be tested with BMO main branch. Releasings and branch maintenance is desribed in BMO [releasing document](https://github.com/metal3-io/baremetal-operator/blob/main/docs/releasing.md)
```

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (Rozzii) [inline design/baremetal-operator/bmo-ci-decoupling.md:27]: Should we wait for the API stabilization work to conclude?

**Review by Rozzii — CHANGES_REQUESTED** [review]

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (Rozzii) [reply design/baremetal-operator/bmo-ci-decoupling.md:27]: https://github.com/metal3-io/metal3-docs/pull/332

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (dtantsur) [reply design/baremetal-operator/bmo-ci-decoupling.md:12]: Let's actually try to get rid of it, even if after ironic-operator is ready.

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (dtantsur) [inline design/baremetal-operator/bmo-ci-decoupling.md:6]: Another benefit is that we may be able to switch the Metal3 job that runs on Ironic to these new tests.

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (kashifest) [reply design/baremetal-operator/bmo-ci-decoupling.md:12]: I agree with @dtantsur

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (kashifest) [reply design/baremetal-operator/bmo-ci-decoupling.md:27]: I think these two work can be done in parallel, no need to wait for API stablization to introduce BMO e2e tests and CI decoupling.

**Review by Rozzii — DISMISSED** [review]

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (tuminoid) [inline design/baremetal-operator/bmo-ci-decoupling.md:23]: Branching out BMO can be technically done, it is just more or less useless until there are some tests in place for the branches so we can merge PRs to them. E2E would be nice to have for one, but to merge E2E in release branch, we need integration tests. This is also too focused on what CAPM3 tests against, while it should detail what BMO 0.x tests against.

From BMO POV I think we should have:
1. release-0.4 branch needs `test-$OS-integration-release-0.4` and `test-$OS-e2e-integration-release-0-4` etc implemented, where BMO 0.4 bracnh is tested against CAPM3 1.5 branch
2. same for release-0.3
3. same for release-0.2

If there are multiple CAPM3 release branches using the same BMO branch, such test should be implemented for all these branches. These should be quite easy to implement via JJB parametrization.

**Review by tuminoid — COMMENTED** [review]
My thoughts on this, two key things to settle:

1. Shall we have integration jobs in BMO branches verifying the release branch compatibility
2. What are the sensible dev-env defaults

In general I feel our CI is too much CAPM3 focused, including our releases, and maybe not in this PR should discuss if dev-env and releases should be focused on higher level Metal3 level, where CAPM3 is just another component same as IPAM or BMO. That would IMO simplify a lot of things in the ecosystem.

File: design/baremetal-operator/bmo-ci-decoupling.md
Comment (lentzi90) [reply design/baremetal-operator/bmo-ci-decoupling.md:27]: I like 3 best. Then I would expect developers to (generally) only modify 1 component at a time and switch that to their own branch.

**Comment by lentzi90** [comment]
> In general I feel our CI is too much CAPM3 focused, including our releases, and maybe not in this PR should discuss if dev-env and releases should be focused on higher level Metal3 level, where CAPM3 is just another component same as IPAM or BMO. That would IMO simplify a lot of things in the ecosystem.

This makes a lot of sense! "Meta" releases would be nice to show what versions of the different components have been tested together.

**Comment by tuminoid** [comment]
Anyways, I've provided my comments as well as most people by now. I'll remove hold. Someone with approver rights to act unless @kashifest wants to change something still. Let's get this merged, and start working on it as it is becoming more and more important to be able to release patch releases from branches as main is getting breaking changes and new features.

/unhold

**Comment by kashifest** [comment]
> Anyways, I've provided my comments as well as most people by now. I'll remove hold. Someone with approver rights to act unless @kashifest wants to change something still. Let's get this merged, and start working on it as it is becoming more and more important to be able to release patch releases from branches as main is getting breaking changes and new features.

I have pushed the typos/fixes you asked. This is ready to go in now. Meanwhile we have already started working on BMO e2e and release branch testing for BMO in parallel. I also agree with the a Metal3 `meta` release process.

**Review by Rozzii — APPROVED** [review]
/approve
This looks fine to me, I don't think there is any issue that has been left uncovered.
The process described in this proposal is already ongoing, branching has happened already, e2e tests are being developed, I don't foresee any deviation from the plan described in this PR.

**Comment by tuminoid** [comment]
/cc @lentzi90 
Just LGTM this and get it merged, it is being half-implemented already, no point having this linger in this state.

**Pull request merged**
